File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/NamedMatcher.java
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern;

import java.util.function.Consumer;
import java.util.function.Function;

public interface NamedMatcher<T extends Matcher> {
    void getNameReferences(Consumer<String> consumer);
    T resolveNames(Function<String, String> resolver);
}

File: common_server/src/main/kotlin/io/github/colintimbarndt/chat_emotes/common/data/unicode/pattern/NamedMatcher.kt
package io.github.colintimbarndt.chat_emotes.common.data.unicode.pattern

/**
 * This is the named variant of [Matcher], meaning that it possibly contains [named references][getNameReferences]
 * to values that need to be resolved first using [resolveNames] in order to use it as a proper [Matcher]
 */
interface NamedMatcher<T : Matcher> {
    fun getNameReferences(): Sequence<String>
    fun resolveNames(resolve: (String) -> String): T
}